crates/core/src/transcript/encoding/encoder.rs

```diff
@@ -1,11 +1,72 @@
-use mpz_circuits::types::ValueType;
-use mpz_core::serialize::CanonicalSerialize;
-use mpz_garble_core::ChaChaEncoder;
+use crate::transcript::{Direction, Subsequence};
+use itybity::ToBits;
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha12Rng;
+use serde::{Deserialize, Serialize};
 
-use crate::transcript::{Direction, Subsequence, RX_TRANSCRIPT_ID, TX_TRANSCRIPT_ID};
+/// Secret used by an encoder to generate encodings.
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct EncoderSecret {
+    seed: Vec<u8>,
+    delta: Vec<u8>,
+}
+
+opaque_debug::implement!(EncoderSecret);
+
+impl EncoderSecret {
+    /// Creates a new secret.
+    ///
+    /// # Arguments
+    ///
+    /// * `seed` - The seed for the PRG.
+    /// * `delta` - Delta for deriving the one-encodings.
+    pub fn new(seed: [u8; 32], delta: [u8; 16]) -> Self {
+        Self {
+            seed: seed.to_vec(),
+            delta: delta.to_vec(),
+        }
+    }
+
+    /// Returns the seed.
+    pub fn to_seed(&self) -> [u8; 32] {
+        self.seed
+            .clone()
+            .try_into()
+            .expect("Seed should be 32 bytes")
+    }
+
+    /// Returns the delta.
+    pub fn to_delta(&self) -> [u8; 16] {
+        self.delta
+            .clone()
+            .try_into()
+            .expect("Delta should be 16 bytes")
+    }
+}
 
-pub(crate) fn new_encoder(seed: [u8; 32]) -> impl Encoder {
-    ChaChaEncoder::new(seed)
+pub(crate) fn new_encoder(secret: &EncoderSecret) -> impl Encoder {
+    ChaChaEncoder::new(secret)
+}
+
+pub(crate) struct ChaChaEncoder {
+    seed: [u8; 32],
+    delta: u128,
+}
+
+impl ChaChaEncoder {
+    pub(crate) fn new(secret: &EncoderSecret) -> Self {
+        let seed = secret.to_seed();
+        let delta = u128::from_le_bytes(secret.to_delta());
+
+        Self { seed, delta }
+    }
+
+    pub(crate) fn new_prg(&self, stream_id: u64) -> ChaCha12Rng {
+        let mut prg = ChaCha12Rng::from_seed(self.seed);
```

Comment:
I don't recall us having a consensus to use chacha12 in prod. I thought we only used chacha**20** in prod. 
Besides, there was recent attacks on 8-round Salsa, https://eprint.iacr.org/2025/289

Reply:
I wasn't aware of our consensus to only use `ChaCha20` in prod. My understanding was that `ChaCha12` is still considered secure (https://github.com/rust-random/rand/issues/932).

Reply:
nvm, we already use blake3 elsewhere in the code which is based on chacha but has only 7 rounds.
So, using chacha12 is justifiable.